src/main/java/net/neoforged/neoforge/common/world/poi/PoiStateSet.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.world.poi;
+
+import com.google.common.collect.Iterators;
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.function.Predicate;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public final class PoiStateSet implements Set<BlockState> {
```

Comment:
Why this custom `Set` instead of just a bog standard `Unmodifyable` one? (Is there even one, unsure?)

Reply:
Because the set needs to be mutable for the extension hook to work while appearing as immutable to the outside and since `PoiType` is a record there is no way (or at least no sane way) to swap out the set after the `PoiType`'s construction.

Reply:
Okey from Discord:
- `PoiType` is a record so we need a custom way to inject custom states.
- This Set type allows package members to additionally insert the custom states.
- Due to the `PoiType` being a record we can not reliably update the field and as such we can not handle this with JVM Native immutable types.